### PR TITLE
[merge to M147] Log GVFS.Service ACL exceptions, and fix upgrade log directory ACL issue

### LINKED ISSUE #767: GVFS.Service does not log failures while attempting to set ACLs
GVFS.Service was not creating the log file until after the attempt to set ACLs. This issue tracks moving the log file creation before setting the ACLs and ensuring any failures during the process are recorded.

### LINKED ISSUE #768: GVFS.Service fails to set ACLs if the directory is not owned by "Administrators"
- Change the owner of `C:\ProgramData\GVFS\GVFS.Upgrade\Logs` to be your user (rather than Administrators) 
- Restart GVFS.Service 
- The service fails to stay up 
- With extra logging (#767): 
 ```
[2019-02-08 12:39:25 -08:00] Information {"Message":"Setting ACLs for C:\\ProgramData\\GVFS\\GVFS.Upgrade\\Logs"} 
[2019-02-08 12:39:25 -08:00] Error {"Area":"GVFSService","Exception":"System.UnauthorizedAccessException: Attempted to perform an unauthorized operation.\r\n   at System.Security.AccessControl.Win32.SetSecurityInfo(ResourceType type, String name, SafeHandle handle, SecurityInfos securityInformation, SecurityIdentifier owner, SecurityIdentifier group, GenericAcl sacl, GenericAcl dacl)\r\n   at System.Security.AccessControl.NativeObjectSecurity.Persist(String name, SafeHandle handle, AccessControlSections includeSections, Object exceptionContext)\r\n   at System.Security.AccessControl.NativeObjectSecurity.Persist(String name, AccessControlSections includeSections, Object exceptionContext)\r\n   at System.Security.AccessControl.FileSystemSecurity.Persist(String fullPath)\r\n   at GVFS.Service.GVFSService.CreateAndConfigureUpgradeLogDirectory() in C:\\Repos\\VFSForGit\\src\\GVFS\\GVFS.Service\\GVFSService.cs:line 433\r\n   at GVFS.Service.GVFSService.CreateAndConfigureProgramDataDirectories() in C:\\Repos\\VFSForGit\\src\\GVFS\\GVFS.Service\\GVFSService.cs:line 399\r\n   at GVFS.Service.GVFSService.OnStart(String[] args) in C:\\Repos\\VFSForGit\\src\\GVFS\\GVFS.Service\\GVFSService.cs:line 165","ErrorMessage":"Unhandled exception in OnStart"}
```

### PR DESCRIPTION
**Pull request by wilbaker**
Fixes #767 and fixes #768

This PR has three commits:

1. Create a tracer before adjusting ACLs so that exceptions (and progress) can be logged
2. Always delete any existing Tools directory before creating a new one
3. There was an issue that if a user previously ran `gvfs upgrade`, `C:\ProgramData\GVFS\GVFS.Upgrade\Logs` would be created with that user as the owner.  Then later, upon upgrading to this version, `Logs` would inherit more restrictive ACLs from its parent directory that would not allow non-elevated `gvfs mount` processes to create new log files.  Attempts to adjust the ACLs on `C:\ProgramData\GVFS\GVFS.Upgrade\Logs` directly by GVFS.Service would fail because `Administrators` was not the owner.  To address this, if GVFS.Service is unable to adjust the ACLs it will create a new `Logs` directory and migrate over any existing log files.

### CODE REVIEW

File: GVFS/GVFS.Service/GvfsService.cs [hunk 1]
```diff
@@ -408,8 +432,73 @@ private void CreateAndConfigureUpgradeLogDirectory()
...
+
+            try
+            {
+                DirectoryInfo tempDirectoryInfo = new DirectoryInfo(tempUpgradeLogsPath);
+                FileInfo[] existingLogFileInfos = tempDirectoryInfo.GetFiles("*", SearchOption.TopDirectoryOnly);
+
+                this.tracer.RelatedInfo($"Moving {existingLogFileInfos.Length} log files from '{tempUpgradeLogsPath}' to '{upgradeLogsPath}'");
+                foreach (FileInfo logFileInfo in existingLogFileInfos)
+                {
+                    File.Move(logFileInfo.FullName, Path.Combine(upgradeLogsPath, logFileInfo.Name));
```
Comment (wilbaker) [inline GVFS/GVFS.Service/GvfsService.cs]: Move over any existing sub-directories as well

File: GVFS/GVFS.Service/GvfsService.cs [hunk 2]
```diff
@@ -155,15 +155,18 @@ protected override void OnStart(string[] args)
                 this.serviceName = serviceName.Substring(ServiceNameArgPrefix.Length);
             }
 
-            this.CreateAndConfigureProgramDataDirectories();
+            string serviceLogsDirectoryPath = Paths.GetServiceLogsPath(this.serviceName);
```
Comment (jamill) [inline GVFS/GVFS.Service/GvfsService.cs]: Maybe add a comment about why the ordering here is important?

File: GVFS/GVFS.Service/GvfsService.cs [hunk 3]
```diff
@@ -408,8 +432,73 @@ private void CreateAndConfigureUpgradeLogDirectory()
...
+
+            this.tracer.RelatedInfo($"Creating new '{upgradeLogsPath}' directory with appropriate ACLs");
+            DirectorySecurity upgradeLogsSecurity = new DirectorySecurity();
+
+            // Protect the access rules from inheritance and remove any inherited rules
+            // (any manually added ACLs are left in place)
+            upgradeLogsSecurity.SetAccessRuleProtection(isProtected: true, preserveInheritance: false);
+
+            // Add ACLs for users and admins
+            WindowsFileSystem.AddUsersAccessRulesToDirectorySecurity(upgradeLogsSecurity, grantUsersModifyPermissions: true);
```
Comment (jamill) [inline GVFS/GVFS.Service/GvfsService.cs]: There is some (non-obvious?) logic here that is pretty specific to adding the access rules to the Logs directory, but it is duplicated. Would there be value in pulling out the common logic so we do not need to ensure that we are adding the ACLs for the upgrade logs directory in 2 places?

File: GVFS/GVFS.Service/GvfsService.cs [hunk 4]
```diff
@@ -374,42 +489,30 @@ private void CreateAndConfigureProgramDataDirectories()
...
+                this.tracer.RelatedInfo($"{nameof(this.GetUpgradeLogsDirectorySecurity)}: '{upgradeLogsPath}' does not exist, creating new ACLs");
                 upgradeLogsSecurity = new DirectorySecurity();
             }
 
             // Protect the access rules from inheritance and remove any inherited rules
-            // (any manually added ACLs are left in place)
             upgradeLogsSecurity.SetAccessRuleProtection(isProtected: true, preserveInheritance: false);
 
-            // Add new ACLs for users and admins
+            // Add new ACLs for users and admins.  Users will be granted write permissions.S
```
Comment (jrbriggs) [inline GVFS/GVFS.Service/GvfsService.cs]: nit: stray S after the .

**Comment by wilbaker** [comment]
This fixes #558, but in my last run I hit a different flaky test (#781).

I am merging this in now (ahead of another run of the full functional tests completing) as we'll be re-running them on the M147 branch before merging it to the shipped branch.

**Pull request merged**